### PR TITLE
#1685 - CTAD for wide

### PR DESCRIPTION
**Pull request by jfalcou**
First part of fixing #1685.
Other deduction guide for other types will come in other PRs.

### CODE REVIEW

File: include/eve/arch/cpu/wide.hpp [hunk 1]
```diff
@@ -991,6 +991,37 @@ namespace eve
...
+  //! @name Deduction Guides
+  //! @{
+  //====================================================================================================================
+
+  /// Allows deduction from a single eve::scalar_value
+  template<scalar_value S> wide(S const&) -> wide<S, expected_cardinal_t<S>>;
+
+  /// Allows deduction from a @callable
+  template<eve::invocable<std::ptrdiff_t, std::ptrdiff_t> Generator>
+  wide(Generator&& g) -> wide<decltype( std::declval<Generator>()(0,0) )>;
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/wide.hpp:1004]: I think this one should be explicit tbh

File: include/eve/arch/cpu/wide.hpp [hunk 2]
```diff
@@ -991,6 +991,37 @@ namespace eve
...
+  template<scalar_value S, std::same_as<S>... Ss>
+  wide(S, Ss...) -> wide<S,fixed<1+sizeof...(Ss)>>;
+
+  /// Allows deduction from a pair of iterators
+  template<std::input_iterator It>
+  wide(It, It) -> wide<std::remove_cvref_t<decltype(*std::declval<It>())>>;
+
+  /// Allows deduction from a Range
+  template<detail::range Range>
+  explicit wide(Range&& r) -> wide<std::remove_cvref_t<decltype(*std::begin(r))>>;
```
Comment (DenisYaroshevskiy) [inline include/eve/arch/cpu/wide.hpp:1020]: I don't like iterator and range ones, I don't think we use them much

File: include/eve/arch/cpu/wide.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/arch/cpu/wide.hpp:1004]: It's a CTAD, there is no notion of explicit.
THe constructor is though

File: include/eve/arch/cpu/wide.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/arch/cpu/wide.hpp:1020]: The constructor exists, I just added the CTAD for completeness-sake.

File: include/eve/arch/cpu/wide.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/arch/cpu/wide.hpp:1004]: oh the ctor is not explicit indeed

Comment (jfalcou) [reply include/eve/arch/cpu/wide.hpp:1004]: Well, writing

```c++
wide w = [](auto i, auto) {....];
```

is cool. Beign explicit will requries () around it.

File: include/eve/arch/cpu/wide.hpp (hunk 2 above)
Comment (DenisYaroshevskiy) [reply include/eve/arch/cpu/wide.hpp:1020]: I meant the ctad here I don't like

File: include/eve/arch/cpu/wide.hpp (hunk 1 above)
Comment (DenisYaroshevskiy) [reply include/eve/arch/cpu/wide.hpp:1004]: Sorry - I meant explicitly specifinh a type

**Review by DenisYaroshevskiy — APPROVED** [review]

**Pull request merged**